zkevm-circuits/src/evm_circuit/op_execution/arithmetic.rs

```diff
@@ -0,0 +1,566 @@
+                    (stack_pointer.clone() - zero)
+                        * (stack_pointer - minus_one),
+                ],
+                lookups: vec![],
+            }
+        };
+
+        let out_of_gas = {
+            let (one, two, three) = (
+                Expression::Constant(F::from_u64(1)),
+                Expression::Constant(F::from_u64(2)),
+                Expression::Constant(F::from_u64(3)),
+            );
+            let (selector, gas_available) = &self.out_of_gas;
+            let gas_overdemand = state_curr.gas_counter.expr() + three.clone()
+                - gas_available.expr();
+            Constraint {
+                name: "AddGadget out of gas",
+                selector: selector.expr(),
+                polys: vec![
+                    (gas_overdemand.clone() - one)
+                        * (gas_overdemand.clone() - two)
+                        * (gas_overdemand - three),
+                ],
+                lookups: vec![],
+            }
+        };
+
+        array::IntoIter::new([success, stack_underflow, out_of_gas])
+            .map(move |mut constraint| {
+                constraint.polys =
+                    [common_polys.clone(), constraint.polys].concat();
+                constraint
+            })
+            .collect()
+    }
+
+    fn assign(
+        &self,
+        region: &mut Region<'_, F>,
+        offset: usize,
+        core_state: &mut CoreStateInstance,
+        execution_step: &ExecutionStep,
+    ) -> Result<(), Error> {
+        match execution_step.case {
+            Case::Success => {
+                self.assign_success(region, offset, core_state, execution_step)
+            }
+            Case::StackUnderflow => {
+                // TODO:
+                unimplemented!()
+            }
+            Case::OutOfGas => {
+                // TODO:
+                unimplemented!()
+            }
+            _ => unreachable!(),
+        }
+    }
+}
+
+impl<F: FieldExt> AddGadget<F> {
+    fn assign_success(
+        &self,
+        region: &mut Region<'_, F>,
+        offset: usize,
+        core_state: &mut CoreStateInstance,
+        execution_step: &ExecutionStep,
+    ) -> Result<(), Error> {
+        core_state.global_counter += 3;
+        core_state.program_counter += 1;
+        core_state.stack_pointer += 1;
+        core_state.gas_counter += 3;
+
+        self.success.swap.assign(
+            region,
+            offset,
+            Some(F::from_u64((execution_step.opcode == OpcodeId::SUB) as u64)),
+        )?;
+        self.success.a.assign(
+            region,
+            offset,
+            Some(execution_step.values[0]),
+        )?;
+        self.success.b.assign(
+            region,
+            offset,
+            Some(execution_step.values[1]),
+        )?;
+        self.success.c.assign(
+            region,
+            offset,
+            Some(execution_step.values[2]),
+        )?;
+        self.success
+            .carry
```

Comment:
Should we calculate the `carry` in the `assign_success` instead of getting it from caller?

Reply:
Make sense, I used to think **auxiliary witness** should be generated in gadget instead of being fed from caller, but it also hard to see which are auxiliary or not before bus mapping has a stable structure. IMO it's ok for now cause it should be trivial to adjust if we need to generate the `carry` here.

Reply:
sounds good.